File: backend/application/ports/itarea_repository.py
# ═══════════════════════════════════════════════════════════════════════════════
# CAPA: APPLICATION — Puertos (Ports / Interfaces)
#
# Analogía: el "contrato laboral" entre el UseCase y el Repository.
# El UseCase dice: "necesito a alguien que sepa crear, listar, actualizar y
# eliminar tareas". Este archivo define ese contrato. No sabe CÓMO se hace —
# solo define QUÉ debe poder hacerse.
#
# Principio: Dependency Inversion (DI) — el UseCase depende de la abstracción
# (este archivo), NO de la implementación concreta (SQLAlchemy, PostgreSQL, etc.).
# Si mañana cambiás SQLite por MongoDB, el UseCase no cambia. Solo cambia
# quien implementa este contrato.
#
# ✅ Puede importar: domain/entities (para conocer la Entity que maneja)
# ❌ NO puede importar: SQLAlchemy, FastAPI, requests, nada externo
# ═══════════════════════════════════════════════════════════════════════════════

from abc import ABC, abstractmethod
# ABC = Abstract Base Class — clase base abstracta de Python estándar.
# Una clase abstracta NO puede instanciarse directamente.
# Solo puede usarse como "plantilla" para que otras clases la implementen.
# Analogía: es como un formulario en blanco — define los campos, pero alguien
# más tiene que llenarlo.

from typing import Optional
# Optional[X] significa que algo puede ser X o None.
# Ej: Optional[Tarea] = puede devolver una Tarea, o None si no existe.

from domain.entities.tarea import Tarea
# Importamos la Entity del Domain. El Puerto CONOCE la Entity porque
# necesita saber qué tipo de objeto maneja.
# Dirección correcta de dependencia: Application → Domain ✅


class ITareaRepository(ABC):
    """
    Puerto (Port) del repositorio de Tareas.

    "I" al inicio = Interface (convención). En Python no hay keyword "interface"
    como en Java o TypeScript; usamos ABC para simularlo.

    Cualquier clase que quiera ser un "repositorio de tareas" DEBE implementar
    todos los métodos marcados con @abstractmethod.
    Si no los implementa, Python lanza un TypeError al intentar instanciarla.
    """

    @abstractmethod
    # "@abstractmethod" marca este método como OBLIGATORIO de implementar.
    # Si olvidás implementarlo en la subclase, Python te avisa con error.
    def crear(self, id: str, titulo: str, completada: bool, creada_en: str) -> Tarea:
        """
        Persiste una nueva Tarea en el almacenamiento y la devuelve completa.

        Parámetros:
          id        — UUID generado antes de llamar (el Domain o UseCase lo genera)
          titulo    — texto de la tarea, ya validado por el Domain
          completada — siempre False en creación
          creada_en — timestamp ISO 8601, generado por el Domain

        Retorna: la Tarea recién creada (ya guardada en DB)
        """
        ...
        # "..." (Ellipsis) es el cuerpo vacío de un método abstracto en Python.
        # Equivalente a "pass" pero semánticamente más claro: "esto es un contrato".

    @abstractmethod
    def listar(self) -> list[Tarea]:
        """
        Devuelve todas las Tareas del almacenamiento.

        list[Tarea] = lista de objetos Tarea (Python 3.9+).
        En versiones anteriores se escribía: List[Tarea] (from typing import List).
        """
        ...

    @abstractmethod
    def obtener_por_id(self, id: str) -> Optional[Tarea]:
        """
        Busca una Tarea por su ID.

        Optional[Tarea] = puede devolver una Tarea o None.
        Si no existe la tarea con ese ID, devuelve None (no lanza error).
        El UseCase decide qué hacer si recibe None (ej: lanzar 404).
        """
        ...

    @abstractmethod
    def actualizar(self, id: str, titulo: str, completada: bool) -> Optional[Tarea]:
        """
        Actualiza los campos de una Tarea existente y la devuelve.

        Devuelve None si la Tarea no existe (ID inválido).
        """
        ...

    @abstractmethod
    def eliminar(self, id: str) -> bool:
        """
        Elimina una Tarea por su ID.

        Devuelve True si la eliminación fue exitosa.
        Devuelve False si la Tarea no existía (no lanza excepción).
        """
        ...

File: frontend/src/application/ports/ITareaRepository.ts
// ═══════════════════════════════════════════════════════════════════════════════
// CAPA: APPLICATION — Puerto (Port / Interface del Repositorio)
//
// Analogía: el "contrato laboral". El UseCase dice:
//   "necesito a alguien que sepa crear, listar, actualizar y eliminar tareas".
// Este archivo define ESE CONTRATO. No sabe CÓMO se implementa.
//
// Principio de Inversión de Dependencias (DIP):
//   Application define QUÉ necesita (este archivo).
//   Infrastructure decide CÓMO hacerlo (TareaRepositoryImpl).
//   El UseCase depende de la ABSTRACCIÓN — nunca de la implementación concreta.
//
// ✅ Puede importar: domain/outputDTO, domain/entities, application/inputDTO (Input DTOs)
// ❌ NO puede importar: fetch, localStorage, ningún detalle de implementación
// ═══════════════════════════════════════════════════════════════════════════════

import type { TareaOutputDTO } from '@/domain/outputDTO/TareaOutputDTO'
// Importamos el OUTPUT DTO — lo que el repositorio promete devolver tras cada operación.
// Vive en domain/outputDTO/ — carpeta separada de entities/ (conceptos distintos).
// El Puerto trabaja con Output DTOs: no con modelos de DB, no con JSON crudo.
// "import type" = TypeScript puro. Desaparece al compilar.

import type { CrearTareaInput, ActualizarTareaInput } from '@/application/inputDTO/TareaInputDTO'
// Importamos los INPUT DTOs — los parámetros que reciben los métodos.


export interface ITareaRepository {
  // "I" al inicio = convención para "Interface". Señaliza que es un contrato.
  // En TypeScript, "interface" describe comportamiento (tiene métodos).
  // Esta interface es TypeScript puro — desaparece al compilar.

  crear(input: CrearTareaInput): Promise<TareaOutputDTO>
  // INPUT DTO entra → OUTPUT DTO sale.
  // "Promise<TareaOutputDTO>" = operación asíncrona que devuelve el Output DTO.
  // El UseCase no sabe si adentro hay HTTP, SQLite, o un mock en memoria.

  listar(): Promise<TareaOutputDTO[]>
  // "TareaOutputDTO[]" = array de Output DTOs. Equivalente a "Array<TareaOutputDTO>".
  // Cuando no hay tareas, devuelve [] (array vacío), no null.

  obtenerPorId(id: string): Promise<TareaOutputDTO | null>
  // "TareaOutputDTO | null" = puede devolver un Output DTO O null.
  // "|" en TypeScript = "union type" — acepta cualquiera de los dos tipos.
  // Si no existe la tarea con ese ID, devuelve null.

  actualizar(id: string, input: ActualizarTareaInput): Promise<TareaOutputDTO>
  // Recibe el ID + Input DTO de actualización → devuelve el Output DTO actualizado.

  eliminar(id: string): Promise<void>
  // "Promise<void>" = no devuelve valor — solo confirma que se eliminó.
}
